extension/persistence/eclipselink/src/main/java/org/apache/polaris/extension/persistence/impl/eclipselink/PolarisSequenceManager.java

```diff
@@ -0,0 +1,115 @@
+
+  /* Get the database platform associated with the `EntityManager` */
+  private static DatabasePlatform getDatabasePlatform(EntityManager session) {
+    EntityManagerImpl entityManagerImpl = session.unwrap(EntityManagerImpl.class);
+    return entityManagerImpl.getDatabaseSession().getPlatform();
+  }
+
+  private static void removeSequence(EntityManager session) {
+    LOGGER.info("Renaming legacy sequence `POLARIS_SEQ` to `POLARIS_SEQ_UNUSED`");
+    String renameSequenceQuery = "ALTER SEQUENCE POLARIS_SEQ RENAME TO POLARIS_SEQ_UNUSED";
+    session.createNativeQuery(renameSequenceQuery).executeUpdate();
+  }
+
+  private static synchronized Optional<Long> getSequenceId(EntityManager session) {
+    DatabasePlatform databasePlatform = getDatabasePlatform(session);
+    if (databasePlatform instanceof PostgreSQLPlatform) {
+      Optional<Long> result = Optional.empty();
+      if (!sequenceCleaned.get()) {
+        try {
+          LOGGER.info("Checking if the sequence POLARIS_SEQ exists");
+          String checkSequenceQuery =
+              "SELECT COUNT(*) FROM information_schema.sequences WHERE sequence_name IN ('polaris_seq', 'POLARIS_SEQ')";
+          int sequenceExists =
+              ((Number) session.createNativeQuery(checkSequenceQuery).getSingleResult()).intValue();
+
+          if (sequenceExists > 0) {
+            LOGGER.info("POLARIS_SEQ exists, calling NEXTVAL");
+            long queryResult =
+                (long) session.createNativeQuery("SELECT NEXTVAL('POLARIS_SEQ')").getSingleResult();
+            result = Optional.of(queryResult);
+          } else {
+            LOGGER.info("POLARIS_SEQ does not exist, skipping NEXTVAL");
+          }
+        } catch (Exception e) {
+          LOGGER.info(
+              "Encountered an exception when checking sequence or calling `NEXTVAL('POLARIS_SEQ')`",
+              e);
+        }
+        if (result.isPresent()) {
+          removeSequence(session);
+        }
+        sequenceCleaned.set(true);
+      }
+      return result;
+    } else {
+      LOGGER.info("Skipping POLARIS_SEQ / NEXTVAL check for platform " + databasePlatform);
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Prepare the `PolarisSequenceManager` to generate IDs. This may run a failing query, so it
+   * should be called for the first time outside the context of a transaction.
+   */
+  public static void initialize(EntityManager session) {
+    // Trigger cleanup of the POLARIS_SEQ if it is present
+    getSequenceId(session);
+  }
+
+  /**
+   * Generates a new ID from `POLARIS_SEQUENCE` or `POLARIS_SEQ` depending on availability. If
+   * `POLARIS_SEQ` exists, it will be renamed to `POLARIS_SEQ_UNUSED`.
+   */
+  public static Long getNewId(EntityManager session) {
+    ModelSequenceId modelSequenceId = new ModelSequenceId();
+
+    // If a legacy sequence ID is present, use that as an override:
+    getSequenceId(session).ifPresent(modelSequenceId::setId);
+
```

Comment:
I feel it is easier to understand if we do this 
```
    ModelSequenceId modelSequenceId = new ModelSequenceId();

    DatabasePlatform databasePlatform = getDatabasePlatform(session);
    if (databasePlatform instanceof PostgreSQLPlatform) {
      getSequenceFromPostgres(session).ifPresent(modelSequenceId::setId);
    }
    
    // Persist the new ID:
    session.persist(modelSequenceId);
    session.flush();
```

Reply:
I think it's better to encapsulate the implementation of `getSequenceId` in that method.